In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import csv

In [2]:
import string
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
import re

## Helpers

In [3]:
def show_na(raw):
    n = len(raw)
    for col in raw.columns:
        na_count = sum(raw[col].isna())
        print("| " + col + " | " + format(na_count, ",d") + " | " + "{:.1%}".format(na_count/n) + " |")

In [4]:
def show_nunique(raw):
    for c in ["country","designation","province","region_1","region_2","variety","winery"]:
        print(c + ": "+str(raw[c].nunique()))

In [5]:
def show_nunique2(raw):
    for c in ["country","designation","province","region_1","region_2","variety","winery","year"]:
        print(c + ": "+str(raw[c].nunique()))

In [6]:
def get_year(t):
    try:
        return [w for w in t.split() if re.search('^[0-9]{4}$', w)][0]
    except:
        return "unknown"

In [7]:
raw = pd.read_csv("winemag-data-130k-v2.csv")

In [8]:
raw["year"] = raw["title"].apply(get_year)

In [9]:
year = raw["year"]

In [10]:
#year.to_csv('years.csv')

In [11]:
show_na(raw)

| Unnamed: 0 | 0 | 0.0% |
| country | 63 | 0.0% |
| description | 0 | 0.0% |
| designation | 37,465 | 28.8% |
| points | 0 | 0.0% |
| price | 8,996 | 6.9% |
| province | 63 | 0.0% |
| region_1 | 21,247 | 16.3% |
| region_2 | 79,460 | 61.1% |
| taster_name | 26,244 | 20.2% |
| taster_twitter_handle | 31,213 | 24.0% |
| title | 0 | 0.0% |
| variety | 1 | 0.0% |
| winery | 0 | 0.0% |
| year | 0 | 0.0% |


| column | unique values | missing count | missing% | type |preprocessing |
| --|--|--|--| -- | -- |
| Unnamed: 0 | 0 | 0.0% | int| deleted |
| country | 43 | 63 | 0.0% | categorical | replace missing with "unknown" -> one-hot encoded |
| description | - | 0 | 0.0% | text | converted into bigram TF-IDF features | 
| designation | 35,777 | 37,465 | 28.8% | categorical |  replace missing with "unknown" -> one-hot encoded  |
| points | -| 0| 0.0% | real | - |
| price | - | 8,996 | 6.9% | real | removed missing row |
| province | 423 | 63 | 0.0% | categorical |  replace missing with "unknown" -> one-hot encoded  |
| region_1 | 1205 | 21,247 | 16.3% | categorical |  replace missing with "unknown" -> one-hot encoded  |
| region_2 | 18 | 79,460 | 61.1% | categorical |  replace missing with "unknown" -> one-hot encoded  |
| taster_name | - | 26,244 | 20.2% | text | deleted |
| taster_twitter_handle | - | 31,213 | 24.0% | text | deleted |
| title | - | 0 | 0.0% | text | extracted new column "year" |
| variety | 698 | 1 | 0.0% |  categorical | removed missing row |
| winery | 15855 | 0 | 0.0% |  categorical | one-hot encoded  |
| year | 89 | 0 | 0.0% |  categorical | new column -> one-hot encoded  

| column | unique values | missing count | missing% |
| --|--|--|--|
| Unnamed: 0 | 0 | 0.0% | int|
| country | 43 | 63 | 0.0% | categorical |
| description | - | 0 | 0.0% | text |
| designation | 35,777 | 37,465 | 28.8% | categorical |
| points | -| 0| 0.0% | real |
| price | - | 8,996 | 6.9% | real |
| province | 423 | 63 | 0.0% | categorical |
| region_1 | 1205 | 21,247 | 16.3% | categorical |
| region_2 | 18 | 79,460 | 61.1% | categorical |
| taster_name | - | 26,244 | 20.2% | text |
| taster_twitter_handle | - | 31,213 | 24.0% | text |
| title | - | 0 | 0.0% | text |
| variety | 698 | 1 | 0.0% |  categorical |
| winery | 15855 | 0 | 0.0% |  categorical |
| year | 89 | 0 | 0.0% |  categorical |

In [12]:
# del raw["Unnamed: 0"]
raw.drop(columns = ["Unnamed: 0","taster_name","taster_twitter_handle","title"], inplace=True)

In [13]:
# raw.info()
show_na(raw)

| country | 63 | 0.0% |
| description | 0 | 0.0% |
| designation | 37,465 | 28.8% |
| points | 0 | 0.0% |
| price | 8,996 | 6.9% |
| province | 63 | 0.0% |
| region_1 | 21,247 | 16.3% |
| region_2 | 79,460 | 61.1% |
| variety | 1 | 0.0% |
| winery | 0 | 0.0% |
| year | 0 | 0.0% |


In [14]:
# raw.dropna(subset = ['price'])

In [15]:
raw.dropna(axis = 0, subset = ["price"],inplace=True)

In [16]:
raw["province"].fillna(value = "UNK_province", inplace=True)
raw["region_1"].fillna(value = "UNK_region_1", inplace=True)
raw["region_2"].fillna(value = "UNK_region_2", inplace=True)
raw["variety"].fillna(value = "UNK_variety", inplace=True)

In [17]:
show_na(raw)

| country | 59 | 0.0% |
| description | 0 | 0.0% |
| designation | 34,779 | 28.7% |
| points | 0 | 0.0% |
| price | 0 | 0.0% |
| province | 0 | 0.0% |
| region_1 | 0 | 0.0% |
| region_2 | 0 | 0.0% |
| variety | 0 | 0.0% |
| winery | 0 | 0.0% |
| year | 0 | 0.0% |


In [18]:
show_nunique2(raw)

country: 42
designation: 35776
province: 423
region_1: 1205
region_2: 18
variety: 698
winery: 15855
year: 89


In [36]:
countries = pd.unique(raw["country"])

#for c in countries:
    #print(c + ": " + str(len(raw[raw["country"] == c])))
    #print(c)
    #print(len(raw[raw["country"] == c]))

In [20]:
US_wines = raw[raw["country"] == "US"]

In [21]:
US_wines

country                                        description  \
2           US  Tart and snappy, the flavors of lime flesh and...   
3           US  Pineapple rind, lemon pith and orange blossom ...   
4           US  Much like the regular bottling from 2012, this...   
10          US  Soft, supple plum envelopes an oaky structure ...   
12          US  Slightly reduced, this wine offers a chalky, t...   
14          US  Building on 150 years and six generations of w...   
19          US  Red fruit aromas pervade on the nose, with cig...   
20          US  Ripe aromas of dark berries mingle with ample ...   
21          US  A sleek mix of tart berry, stem and herb, alon...   
23          US  This wine from the Geneseo district offers aro...   
25          US  Oak and earth intermingle around robust aromas...   
29          US  Clarksburg is becoming a haven for Chenin Blan...   
33          US  Rustic and dry, this has flavors of berries, c...   
34          US  This shows a tart, green gooseberry flavor tha...   
35          US  As with many of the Erath 2010 vineyard design...   
41          US  A stiff, tannic wine, this slowly opens and br...   
43          US  The clean, brisk mouthfeel gives this slightly...   
45          US  Right out of the starting blocks this is an oa...   
47          US  This is a sweet wine with flavors of white sug...   
48          US  This bottling resembles the New Zealand paradi...   
55          US  This shows jelly-like flavors of orange and pe...   
56          US  This is weighty, creamy and medium to full in ...   
59          US  Aromas of cranberry, barrel spice and herb are...   
60          US  Syrupy and dense, this wine is jammy in plum a...   
62          US  The aromas are brooding, with notes of barrel ...   
64          US  There are intriguing touches to the nose of th...   
67          US  A blend of Merlot and Cabernet Franc, this win...   
68          US  Very deep in color and spicy-smoky in flavor, ...   
70          US  Aromas of vanilla, char and toast lead to ligh...   
71          US  Big oak defines this robustly dense and extrac...   
...        ...                                                ...   
129880      US  Apple blossom intrigues on the nose of this li...   
129882      US  This wine is mostly Cabernet Sauvignon (78%), ...   
129885      US  Lemon-lime soda aromas meet with orange blosso...   
129888      US  Round red cherry, deep plum and ripe cranberry...   
129891      US  Chocolate, mocha and coconut notes highlight t...   
129895      US  A blend of vineyards including the winery's Mi...   
129896      US  Smoothly seductive in texture and body, this m...   
129897      US  This hearty, rich bottling needs time to breat...   
129898      US  Plum and cherry fruit show on the nose of this...   
129899      US  Aged entirely in stainless steel, this combine...   
129904      US  A blend of Sauvignon Blanc, Sauvignon Gris and...   
129906      US  Pencil shaving and graphite notes lend a class...   
129909      US  Hailing from an extreme coastal vineyard, this...   
129912      US  From the Ranch House block of the famous viney...   
129913      US  This is taut and dense, and requires time and ...   
129914      US  Fritz has made tremendous progress with Cab ov...   
129915      US  The third vintage of Mith is the best yet, com...   
129919      US  This ripe, rich, almost decadently thick wine ...   
129920      US  Shows the clean, citrus acid juiciness and sil...   
129926      US  This pure Syrah from Reininger's estate vineya...   
129927      US  It's 100% Cabernet Franc, a delightful, bright...   
129935      US  This has lovely raspberry and cranberry fruit,...   
129940      US  This is the winery's bells and whistles Chardo...   
129941      US  A Chardonnay with an unusual companion, 8% Sém...   
129942      US  This is classic in herbaceous aromas and flavo...   
129945      US  Hailing from one of the more popular vineyards...   
12994

In [69]:
US_wines.to_csv('USwines.csv')

# output csv

In [ ]:
raw.drop(columns = ["title","description"], inplace=True)

In [ ]:
raw.to_csv('wines.csv')

In [ ]:
raw.shape

In [ ]:
#raw = raw.iloc[1:20000,:]

In [ ]:
raw = raw[raw["price"]<300]

In [ ]:
show_nunique2(raw)

In [ ]:
country = pd.get_dummies(raw["country"])
designation = pd.get_dummies(raw["designation"])
province = pd.get_dummies(raw["province"])
region_1 = pd.get_dummies(raw["region_1"])
region_2 = pd.get_dummies(raw["region_2"])
variety = pd.get_dummies(raw["variety"])
winery = pd.get_dummies(raw["winery"])
points = pd.DataFrame(raw["points"])
price = pd.DataFrame(raw["price"])

year = pd.get_dummies(raw["year"])

In [ ]:
from sklearn.preprocessing import OneHotEncoder
onehotencoder = OneHotEncoder()

In [ ]:
X_cat = onehotencoder.fit_transform(raw[["region_1","variety","year","winery"]]).toarray()

### Description: TF-IDF

In [22]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [23]:
from sklearn.feature_extraction import text

In [24]:
#text.ENGLISH_STOP_WORDS

In [25]:
vectorizer = TfidfVectorizer(ngram_range = (2,2),
                            stop_words = 'english',
                             lowercase = True)
# max_features
# norm = l2 by default

In [26]:
desc = vectorizer.fit_transform(US_wines["description"])

In [27]:
len(vectorizer.vocabulary_)

423314

In [28]:
tf_idf = (desc.toarray()) 
tf_idf.shape

(54265, 423314)

In [ ]:
# print("\n\nScores : \n", tf_idf) 

In [ ]:
type(tf_idf)

In [18]:
TFIDF = pd.DataFrame(tf_idf)

In [19]:
TFIDF.shape

(120975, 738234)

In [20]:
TFIDF.head(1)

0       1       2       3       4       5       6       7       8       \
0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   9       ...  738224  738225  738226  738227  738228  738229  738230  \
0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   738231  738232  738233  
0     0.0     0.0     0.0  

[1 rows x 738234 columns]

In [21]:
TFIDF.to_csv('TFIDF.csv')

KeyboardInterrupt: 